### ISSUE TITLE
Excessive memory usage and processing time when training with 1000 annotations per image.

### ISSUE THREAD

**Issue by VdLMV** [ultralytics/ultralytics#492]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

See code below and set the variable annot:
- Number of annotations  is 10: gpu mem = 1.6Gb, execution time ~1 s.
- Number of annotations  is 1000: gpu mem = 21.7Gb, execution time ~10 s.

### Environment

YOLOV8 version "8.0.9" torch 11.1 cuda 11.7 (RTX3090)
Ubuntu 20.04
Python 3.10

### Minimal Reproducible Example

from ultralytics import YOLO
from ultralytics.yolo.v8.detect.train import Loss
import torch
device = "cuda:0"
model = YOLO('yolov8s.pt').model.to(device)
loss = Loss(model)
imgsz, bs, annot = 640, 16, 10
img = torch.ones((bs, 3, imgsz, imgsz))
idx = torch.ones((bs * annot))
for b in range(bs): idx[b*annot: (b+1)*annot] = b
cls = torch.ones((bs * annot, 1))
bboxes = torch.ones((bs * annot, 4))
bboxes[:, 0:2] = torch.rand(2) / 2.
bboxes[:, 2:4] = torch.rand(2) / 3.
batch = dict(img=img, batch_idx=idx, cls=cls, bboxes=bboxes)
for _ in range(100):
    with torch.cuda.amp.autocast():
        pred = model(batch['img'].to(device))
        _ = loss(pred, batch)

### Additional

In order to make a minimal example I had to change in train.py:
        # loss[0] *= self.hyp.box  # box gain
        # loss[1] *= self.hyp.cls  # cls gain
        # loss[2] *= self.hyp.dfl  # dfl gain
        loss[0] *= self.hyp['box']  # box gain
        loss[1] *= self.hyp['cls']  # cls gain
        loss[2] *= self.hyp['dfl']  # dfl gain

The following might be related or unrelated to the bug report above:
Although  I can run and debug the minimal example above using PyCharm with remote interpreter, it is not possible to debug the 'standard' traing pipeline with the code below:

from ultralytics import YOLO
model = YOLO('yolov8s.pt')  # load a pretrained YOLOv8n detection model
model.train(data='coco128.yaml', epochs=4)  # train the model
res = model.predict(source='https://ultralytics.com/images/bus.jpg')  # predict on an image

When setting a breakpoint in the Loss class the debug process is terminated with "Process finished with exit code 139".
Running (not debugging) the code from Pycharm works as expected.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by VdLMV** [comment]
@AyushExel Thx for your reaction, but I don't understand your explanation.
Maybe there is some confusion. The issue is the number of annotations per image. In yolov5 there was not a noticeble increase in gpu memory usage and training time between training datasets with only a few annotations per image and datasets with many, like 1000, annotations per image. The amount of memory needed for the annotations is relative small compared to memory needed for the images.
When debugging the minimal example there is an unexepected huge increase (1.4 -> 21Gb) of gpu memory the first time the Loss function (TaskAlignedAssigner.get_pos_mask ) is called.

**Comment by Laughing-q** [comment]
@VdLMV yeah the TAL assigner is not as efficient as yolov5 when handling the growth of instance number.
Let's say there're 3 images which the number of labels of each are 1, 3 and 50, in yolov5 the shape will be (1+3+50, 4) when assigning labels, but in v8 the shape will be (3, 50, 4) which is (bs, the max_num of instance, 4).

Also we tried to optimize the cuda memory, but we found the only way is to use for loop while assigning labels, but it's too slow...

We chose TAL as it's more accurate, meanwhile the increase of cuda memory is acceptable as usually we don't have such gap of instancs number. i.e 10 and 1000.
